### PR TITLE
Clarify routing logic

### PR DESCRIPTION
**Pull request by olix0r**
The relationship between protocol & authority was subtle.

This change mostly adds documentation, though the `Destination` enum has
been renamed to `Dst` with `Bound` and `Unbound` variants, replacing
`LocalSvc` and `External`, which had misleading connotations.

Protocol::into_key has been removed so that the reuse logic is explicit
and readable.

Based on #555

### CODE REVIEW

File: proxy/src/bind.rs [hunk 1]
```diff
@@ -322,19 +321,4 @@ impl Protocol {
...
     }
-
-    pub fn is_cachable(&self) -> bool {
-        match *self {
-            Protocol::Http2 | Protocol::Http1(Host::Authority(_)) => true,
-            _ => false,
-        }
-    }
-
-    pub fn into_key<T>(self, key: T) -> Reuse<(T, Protocol)> {
```
Comment (hawkw) [inline proxy/src/bind.rs]: I think removing this logic from `Bind` is a good call. At the time, it felt like a good reduction in code duplication, but looking back, it definitely introduces some unnecessary complexity. 👍

File: proxy/src/outbound.rs [hunk 2]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
+        let qualified_name = req.uri().authority_part()
             .cloned()
-        // Therefore, we need to check the host header as well as the URI
-        // for a valid authority, before we fall back to SO_ORIGINAL_DST.
             .or_else(|| h1::authority_from_host(req))
             .map(|authority| {
-                FullyQualifiedAuthority::normalize(
-                    &authority,
-                    &self.default_namespace)
+                FullyQualifiedAuthority::normalize(&authority, &self.default_namespace)
```
Comment (briansmith) [inline proxy/src/outbound.rs]: `h1::authority_from_hist(req)` does the `SO_ORIGINAL_DST` logic itself, so if it returns None then we can return None right away. That is, we just did the "If there is no original dst, then we have nothing!" part from below, and we don't need to do SO_ORIGINAL_DST stuff below, right?

File: proxy/src/fully_qualified_authority.rs [hunk 3]
```diff
@@ -41,13 +41,14 @@ impl FullyQualifiedAuthority {
             }
         };
 
-        // parts should have a maximum 4 of pieces (name, namespace, svc, zone)
+        // Detect local service names in the form _name.namespace.svc.zone_. _zone_ may
+        // contain dotted labels.
```
Comment (hawkw) [inline proxy/src/fully_qualified_authority.rs]: These comments are a lot clearer. I'd really like to see this function refactored a bit more, as I found it pretty hard to follow (especially the reliance on early returns) but that's probably out of scope for this PR.

File: proxy/src/outbound.rs [hunk 4]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
-    LocalSvc(FullyQualifiedAuthority),
-    External(SocketAddr),
-}
-
-impl From<FullyQualifiedAuthority> for Destination {
-    #[inline]
-    fn from(authority: FullyQualifiedAuthority) -> Self {
-        Destination::LocalSvc(authority)
-    }
+pub enum Dst {
```
Comment (hawkw) [inline proxy/src/outbound.rs]: This is minor and I could be totally out of line, but personally i'm not sure if an `enum` where each variant owns a `Protocol` is any clearer than a tuple of (`enum`, `Protocol`)...

File: proxy/src/outbound.rs [hunk 5]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
+    /// Determines the destination service for a given Request.
+    ///
+    /// Destinations are either bound or unbound. Unbound destinations must be resolved
+    /// through the controller to produce a list of endpoints. Bound destinations use the
+    /// client-provided destination address.
+    ///
+    /// For HTTP/2 requests, a service may support several :authority values. HTTP/1
+    /// requests, on the other hand, may not share a service across authorities.
+    ///
+    /// Services may not be reused if they do not include an authority.
```
Comment (hawkw) [inline proxy/src/outbound.rs]: Comments in this function are significantly better now --- this is great!

File: proxy/src/outbound.rs [hunk 6]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
-        // If we can't fully qualify the authority as a local service,
-        // and there is no original dst, then we have nothing! In that
-        // case, we return `None`, which results an "unrecognized" error.
+        // If the request has an authority that looks like something the destination
+        // service might be able to resolve, treat the name as unbound. Note that this
+        // will result in different keys for different original authority values, so
+        // requests for `foo` and `foo.ns` won't be satisfied by the same connections.
+        if let Some(qn) = qualified_name {
+            if qn.use_destination_service {
+                return Some(Reuse::Reusable(Dst::Unbound(qn.name, proto)));
```
Comment (hawkw) [inline proxy/src/outbound.rs]: This may be an unpopular opinion, but I find that early returns tend to make functions like this less clear.

File: proxy/src/outbound.rs (hunk 2 above)
Comment (olix0r) [reply proxy/src/outbound.rs]: I don't think that's right.

`h1::normalize_our_view_of_uri` does `SO_ORIGINAL_DST` logic:
https://github.com/runconduit/conduit/blob/0fbcf956321b5ef1295319fa1d95d4694777038b/proxy/src/transparency/h1.rs#L20-L36

`h1::authority_from_host` does not:
https://github.com/runconduit/conduit/blob/0fbcf956321b5ef1295319fa1d95d4694777038b/proxy/src/transparency/h1.rs#L47-L60

File: proxy/src/outbound.rs [hunk 7]
```diff
@@ -139,19 +136,18 @@ where
     /// changed.
     fn bind_service(
         &mut self,
-        key: &Self::Key,
+        dst: &Self::Key,
     ) -> Result<Self::Service, Self::RouteError> {
-        let &(ref dest, ref protocol) = key;
-        debug!("building outbound {:?} client to {:?}", protocol, dest);
+        debug!("building outbound client to {:?}", dst);
```
Comment (hawkw) [inline proxy/src/outbound.rs]: TIOLI, really minor nit, but I think the original log line, which would say something like "building outbound Http1 client to 127.0.0.1:80" was probably more readable than the new log line, which might say things like "building outbound Http1 client to Dst::Unbound("127.0.0.1:80", Http1)". This probably doesn't really matter at all since it's a debug log line...

File: proxy/src/fully_qualified_authority.rs (hunk 3 above)
Comment (briansmith) [reply proxy/src/fully_qualified_authority.rs]: This is all going to get replaced very soon with stuff that is much clearer.

File: proxy/src/outbound.rs (hunk 6 above)
Comment (olix0r) [reply proxy/src/outbound.rs]: Moving logic into else clauses would make this even more convoluted. Early returns are a necessary evil.

File: proxy/src/outbound.rs (hunk 4 above)
Comment (olix0r) [reply proxy/src/outbound.rs]: The intent is that a `Dst` is meaningless without a protocol.  It doesn't make sense (to me) for `Dst` _not_ to have a protocol.

Comment (hawkw) [reply proxy/src/outbound.rs]: That makes sense. I guess my instinct is to not duplicate fields across enum variants when they're the same?

Comment (briansmith) [reply proxy/src/outbound.rs]: I agree that Dst is meaningless without a protocol. However, the current structure actually implies that we might add new variants that don't have a protocol since whether there's a protocol field is determined per-variant. If every Dst has a protocol then it should be like @hawkw suggested:
```rust
pub struct Dst {
    protocol: Protocol,
    binding: DstBinding,
}
enum DstBnding {
    Unbound(FullyQualifiedAuthority),
    Bound(SocketAddr),
}
```

(Just like `bind::BindProtocol`.)

File: proxy/src/outbound.rs [hunk 8]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
     >>>>;
 
+    /// Determines the destination service for a given Request.
+    ///
+    /// Destinations are either bound or unbound. Unbound destinations must be resolved
+    /// through the controller to produce a list of endpoints. Bound destinations use the
+    /// client-provided destination address.
+    ///
+    /// For HTTP/2 requests, a service may support several :authority values. HTTP/1
+    /// requests, on the other hand, may not share a service across authorities.
```
Comment (briansmith) [inline proxy/src/outbound.rs]: I suggest: "HTTP/2 connections may multiplex connections for multiple authorities (as identified by `:authority`) on the same connection HTTP/1 connections must have one authority per connection." In particular, this isn't a per-service thing, but a per-connection thing. Also, please move this comment to be with the code that implements this logic. (See my note below.)

File: proxy/src/outbound.rs [hunk 9]
```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
...
+        // service might be able to resolve, treat the name as unbound. Note that this
+        // will result in different keys for different original authority values, so
+        // requests for `foo` and `foo.ns` won't be satisfied by the same connections.
+        if let Some(qn) = qualified_name {
+            if qn.use_destination_service {
+                return Some(Reuse::Reusable(Dst::Unbound(qn.name, proto)));
+            }
+        }
+
+        // Otherwise, send the request to the client-bound address.
```
Comment (briansmith) [inline proxy/src/outbound.rs]: The logic from here down is the same as the similar logic for `Inbound`. This code is better documented, so let's factor this version of the code out into a function that `Inbound` and `Outbound` share, with the addition comment I suggested ("HTTP/2 connections may multiplex connections for multiple authorities (as identified by `:authority`) on the same connection HTTP/1 connections must have one authority per connection.").

Comment (hawkw) [reply proxy/src/outbound.rs]: I also wanted to reuse this code between `Inbound` and `Outbound` (which was the motivation behind the rather poorly thought out `Protocol::into_key` function), but note that `Inbound` also has a default address to fall back to, while `Outbound` does not.

File: proxy/src/outbound.rs (hunk 8 above)
Comment (olix0r) [reply proxy/src/outbound.rs]: I think it actually _is_ about services and not connections:

the router contract is that Recognize produces a key that is used to bind a service. this service may maintain an arbitrary number of connections.  we care that we have a service for each unique authority.

File: proxy/src/outbound.rs (hunk 9 above)
Comment (briansmith) [reply proxy/src/outbound.rs]: `default_addr: Option<SocketAddr>` so `Outbound` should be able to just pass `None` for it.
